src/main/java/io/r2dbc/proxy/callback/ProxyFactory.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.r2dbc.proxy.callback;
+
+import io.r2dbc.proxy.core.ConnectionInfo;
+import io.r2dbc.proxy.core.QueryExecutionInfo;
+import io.r2dbc.spi.Batch;
+import io.r2dbc.spi.Connection;
+import io.r2dbc.spi.ConnectionFactory;
+import io.r2dbc.spi.Result;
+import io.r2dbc.spi.Statement;
+
+/**
+ * @author Tadaya Tsuyukubo
+ */
+public interface ProxyFactory {
+
+    void setProxyConfig(ProxyConfig proxyConfig);
+
+    ConnectionFactory createConnectionFactory(ConnectionFactory connectionFactory);
+
+    Connection createConnection(Connection connection, ConnectionInfo connectionInfo);
+
+    Batch<?> createBatch(Batch<?> batch, ConnectionInfo connectionInfo);
+
+    Statement<?> createStatement(Statement<?> statement, String query, ConnectionInfo connectionInfo);
+
+    Result createResult(Result result, QueryExecutionInfo executionInfo);
+
```

Comment:
Some of these methods sound strange. `ConnectionFactory createConnectionFactory(ConnectionFactory connectionFactory)` for example. `createConnectionFactory` implies it's creating a new one, but it has one as an input. Does this wrap one with another?

Reply:
Yes, the return value is the wrapped(proxied) input value.
Probably, rename each method to `createProxy...` for better naming??

Example usage would be:
```java
Connection proxiedConnecition = proxyFactory.createProxyConnection(originalConnection, connectionInfo);
```

Reply:
Much better.

Reply:
Renamed to `createProxy...`